### ISSUE TITLE
Why triplet loss is limited by 0?

### ISSUE THREAD

**Issue by truonghaophu** [davidsandberg/facenet#811]
Hi everyone,
Can someone explain why the triplet loss is limited by 0? Why triplet loss can't be negative?
```python
with tf.variable_scope('triplet_loss'):
        pos_dist = tf.reduce_sum(tf.square(tf.subtract(anchor, positive)), 1)
        neg_dist = tf.reduce_sum(tf.square(tf.subtract(anchor, negative)), 1)
        
        basic_loss = tf.add(tf.subtract(pos_dist,neg_dist), alpha)
        loss = tf.reduce_mean(tf.maximum(basic_loss, 0.0), 0)
```
Any help would be appreciated.

**Comment by kanikel** [comment]
apparently, there is a tf.maximum(..,0) in the last line, if loss is negative, loss will be 0

**Comment by truonghaophu** [comment]
@kanikel Thanks for your replied. I already understand the code above. Things I didn't understand is in the [paper](https://arxiv.org/pdf/1503.03832.pdf), the author mention the triplet loss like below:

![image](https://user-images.githubusercontent.com/13709414/42668443-e81e30c4-867a-11e8-9d2a-79f3d72fd61a.png)

I wonder can the triplet loss be **negative**?

**Comment by davidsandberg** [comment]
Hi,
The operator [...]+ that you can see in the equation does exactly that.
If the loss is 0 it means that you have found an optimal solution, and adjusting the model parameters in that case doesn't make sense.

**Comment by truonghaophu** [comment]
@davidsandberg Hi David, thanks for your replied.

> If the loss is 0 it means that you have found an optimal solution, and adjusting the model parameters in that case doesn't make sense.

Yes, what you said is totally true. But when the loss going down to 0. If I change the code to:
```python
loss = tf.reduce_mean(basic_loss, 0)
```
and the training is continue with negative loss like normal. My question is why not let the loss be negative but limit it by 0? Is there any reason behind that?

**Comment by Adarine99** [comment]
I think the key is `--alpha`,it means 'Positive to negative triplet distance margin',so limit the loss to 0

**Comment by Vozf** [comment]
So have you found an answer?

**Comment by rasbt** [comment]
> and the training is continue with negative loss like normal. My question is why not let the loss be negative but limit it by 0?

you would never stop updating then. The best case scenario is d(A, P)=0 and d(A, N)=4 hence L=0 - 4 + alpha = -3.5 (if alpha = 0.5). You don't want to update in such cases because this is pretty good :). Thus, to prevent updating/backprop signals if the anchor-negative distance is  large, you set the loss to 0 in such cases
